modules/dnn/src/layers/convolution_layer.cpp

```diff
@@ -1210,6 +1218,158 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
+        else if (padMode == "SAME")
+        {
+            /* TensorFlow Logic:
+             * total_padding[i] = (o[i] - 1) * s[i] + effective_k[i] - i[i]
+             *
+             * if total padding is odd, the input is padded towards the end
+             */
+            std::vector<int> inShape(std::begin(input_shape) + 2, std::end(input_shape)), outShape;
+            getConvPoolOutParams(inShape, kernel_size, strides, padMode, dilations, outShape);
+
+            for (int i = 2; i < rank; i++)
+            {
+                const auto j = i - 2; /* filter index */
+                const auto effective_kernel_size = dilations[j] * (kernel_size[j] - 1) + 1;
+                const auto required_total_padding =
+                    std::max<int>(0, (outShape[j] - 1) * strides[j] + effective_kernel_size - inShape[j]);
+
+                common_padding[i] = required_total_padding / 2;
+                padding_left[i] = 0;
+                padding_right[i] = required_total_padding % 2;
+            }
+        }
+        else
+        {
+            CV_Error(Error::StsNotImplemented, "Specified padding mode not supported by ConvolutionLayer");
+        }
+
+        /* csl::Convolution supports symmetric padding only; hence, we deal with asymmetric padding by
+         * copying the input to a bigger tensor and pad the sides manually
+         */
+        for (int i = 0; i < rank; i++)
+            input_shape[i] += padding_left[i] + padding_right[i];
+
+        /* if the actual input shape and the new input shape do not match; we need to transform the input */
+        transform_required = input_shape != input_wrapper->getShape();
+        if (transform_required)
+        {
+            transformedInput.resize(std::begin(input_shape), std::end(input_shape));
+            inputTransformer = csl::TensorTransform<float>(cudnnHandle, padding_left, padding_right);
+        }
+
+        csl::Convolution<float>::params_type params;
+
+        auto& ishape = params.input_shape;
+        ishape.assign(std::begin(input_shape), std::end(input_shape));
+
+        auto& fshape = params.filter_shape;
+        fshape.resize(ishape.size());
+        fshape[0] = output_feature_maps;
+        fshape[1] = input_feature_maps_per_group;
+        std::copy(std::begin(kernel_size), std::end(kernel_size), std::begin(fshape) + 2);
+        CV_Assert(fshape.size() == kernel_size.size() + 2);
+
+        params.padding.assign(std::begin(common_padding) + 2, std::end(common_padding));
+        params.stride = strides;
+        params.dialation = dilations;
+        params.groups = groups;
+
+        convoluter = csl::Convolution<float>(cudnnHandle, params);
+        scratch_mem_in_bytes = convoluter.get_workspace_size();
+    }
+
+    csl::cudnn::Handle cudnnHandle;
+    csl::Tensor<float> filtersTensor, biasTensor;
+    csl::Convolution<float> convoluter;
+
+    bool transform_required;
+    csl::Tensor<float> transformedInput;
```

Comment:
This allocates memory for a tensor which can be only be used by this layer. The correct way would be to request extra memory in `initCUDA` and take a block from the workspace during the forward pass.

Currently, there is no mechanism for splitting the workspace memory without manually playing with pointers. I'll come up with something neat for dealing with workspaces soon. Until then, `transformedInput` will be a member.